syntax = "proto3";

package smartfactory;

option go_package = "github.com/smartfactory/proto;proto";

service DetectionService {
    rpc Detect(DetectRequest) returns (DetectResponse);
    rpc StreamDetect(stream DetectRequest) returns (stream DetectResponse);
    rpc HealthCheck(Empty) returns (HealthResponse);
}

message Empty {}

message DetectRequest {
    bytes image_data = 1;
    string camera_id = 2;
    int64 timestamp = 3;
    float confidence_threshold = 4;
}

message DetectResponse {
    repeated Detection detections = 1;
    SafetyCheck safety_check = 2;
    repeated ZoneViolation zone_violations = 3;
    float processing_time_ms = 4;
    int64 timestamp = 5;
    string camera_id = 6;
}

message Detection {
    int32 class_id = 1;
    string class_name = 2;
    float confidence = 3;
    BoundingBox bbox = 4;
    int32 track_id = 5;
}

message BoundingBox {
    int32 x1 = 1;
    int32 y1 = 2;
    int32 x2 = 3;
    int32 y2 = 4;
}

message SafetyCheck {
    bool has_violations = 1;
    repeated string violations = 2;
    int32 people_count = 3;
    int32 violation_count = 4;
    int32 compliant_count = 5;
    float compliance_rate = 6;
}

message ZoneViolation {
    string zone_id = 1;
    string zone_name = 2;
    string severity = 3;
    int32 person_track_id = 4;
    repeated string missing_ppe = 5;
    int64 timestamp = 6;
}

message HealthResponse {
    string status = 1;
    bool model_loaded = 2;
    string model_type = 3;
    string device = 4;
}

message Zone {
    string id = 1;
    string name = 2;
    string severity = 3;
    repeated Point polygon = 4;
    repeated string required_ppe = 5;
    bool enabled = 6;
}

message Point {
    float x = 1;
    float y = 2;
}
